File: benchmarks/bench.py
import time

def sieve(n):
    is_prime = bytearray([1]) * (n + 1)
    is_prime[0] = 0
    is_prime[1] = 0
    i = 2
    while i * i <= n:
        if is_prime[i]:
            j = i * i
            while j <= n:
                is_prime[j] = 0
                j += i
        i += 1
    return sum(is_prime)

def fib_count():
    U64_MAX = (1 << 64) - 1
    a, b = 0, 1
    count = 0
    while True:
        nxt = a + b
        if nxt > U64_MAX:
            break
        a, b = b, nxt
        count += 1
        if b == U64_MAX:
            break
    return count

N = 10_000_000

t0 = time.perf_counter()
count = sieve(N)
elapsed_sieve = (time.perf_counter() - t0) * 1000
print(f"Primes up to {N}: {count} ({elapsed_sieve:.1f} ms)")

t1 = time.perf_counter()
fib_n = fib_count()
elapsed_fib = (time.perf_counter() - t1) * 1000
print(f"Fibonacci numbers under u64: {fib_n} ({elapsed_fib:.3f} ms)")

File: benchmarks/bench.js
// Node.js benchmark

function sieve(n) {
  const isPrime = new Uint8Array(n + 1).fill(1);
  isPrime[0] = 0;
  isPrime[1] = 0;
  for (let i = 2; i * i <= n; i++) {
    if (isPrime[i]) {
      for (let j = i * i; j <= n; j += i) {
        isPrime[j] = 0;
      }
    }
  }
  let count = 0;
  for (let i = 2; i <= n; i++) {
    if (isPrime[i]) count++;
  }
  return count;
}

function fibCount() {
  const U64_MAX = 18446744073709551615n;
  let a = 0n, b = 1n;
  let count = 0;
  while (true) {
    const next = a + b;
    if (next > U64_MAX) break;
    a = b;
    b = next;
    count++;
    if (b === U64_MAX) break;
  }
  return count;
}

const N = 10_000_000;

const t0 = performance.now();
const count = sieve(N);
const elapsedSieve = (performance.now() - t0).toFixed(1);
console.log(`Primes up to ${N}: ${count} (${elapsedSieve} ms)`);

const t1 = performance.now();
const fibN = fibCount();
const elapsedFib = (performance.now() - t1).toFixed(3);
console.log(`Fibonacci numbers under u64: ${fibN} (${elapsedFib} ms)`);
